File: [D]binary-search/[D]binary-search/74-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows, cols = len(matrix), len(matrix[0])

        def get_num(idx):
            r, c = divmod(idx, cols)
            return matrix[r][c]

        left, right, boundary = 0, rows * cols - 1, - 1
        while left <= right:
            m = (left + right) // 2
            num = get_num(m)
            if num == target:
                boundary = m
                break
            elif num > target:
                right = m - 1
            else:
                left = m + 1
        return boundary != - 1
    
# time O(log(mn))
# space O(1)
# using binary search and search in a sorted array for specific val

File: [D]binary-search/[D]binary-search/74-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int left = 0;
        int right = rows * cols - 1;
        while (left <= right) {
            int m = left + (right - left) / 2;
            int r = m / cols;
            int c = m % cols;
            if (matrix[r][c] > target) {
                right = m - 1;
            } else if (matrix[r][c] == target) {
                return true;
            } else {
                left = m + 1;
            }
        }
        return false;
    }
}

// time O(log(mn))
// space O(1)
// using binary search and search in a sorted array for specific val